### PR TITLE
Mm 445 - Filter out "<details>...</details>" text from PR descriptions fixes #445

### LINKED ISSUE #445: Filter out "<details>...</details>" text from PR descriptions
Bots such as @dependabot often make posts structured with the [`<details>`](https://gist.github.com/ericclemmons/b146fe5da72ca1f706b2ef72a20ac39d) tag to hide details that can be expanded when clicked. When this is rendered in a Mattermost post, the result shown is the raw HTML containing the details. We can hide this text since it is mostly unnecessary information that is not properly formatted.

The `<details>` block can be replaced with the text "(details collapsed)"

image

_Issue created from a [Mattermost message](https://community-daily.mattermost.com/_redirect/pl/c6asq8azj3gz8et78kge9hq6ea) *by `@thiefmaster`*._

### PR DESCRIPTION
**Pull request by sibasankarnayak**
replaced the text <details>...</details> with details collapsed

ticket here
Fixes https://github.com/mattermost/mattermost-plugin-github/issues/445

### CODE REVIEW

File: server/plugin/webhook.go [hunk 1]
```diff
@@ -270,6 +272,10 @@ func (p *Plugin) postPullRequestEvent(event *github.PullRequestEvent) {
 		}
 
 		if action == "opened" {
+			policy := bluemonday.StrictPolicy()
+			policy.SkipElementsContent("details")
+			newPRMessage = strings.ReplaceAll(newPRMessage, "<details>", "details collapsed.<details>")
+			newPRMessage = policy.Sanitize(newPRMessage)
```
Comment (mickmister) [inline server/plugin/webhook.go:278]: Great solution! I have one suggestion to change the substitution string

```suggestion
			newPRMessage = strings.ReplaceAll(newPRMessage, "<details>", "`(details collapsed)`<details>")
			newPRMessage = policy.Sanitize(newPRMessage)
```

**Review by mickmister — CHANGES_REQUESTED** [review]
@sibasankar-demansol Great work! I have one request regarding the new text that is shown

**Comment by mickmister** [comment]
I wonder if we should also apply the substitution on other entities such as issues and comments. @hanzei What do you think? This is really being made to solve the `dependabot` scenario which is solved by this PR.

File: server/plugin/webhook.go (hunk 1 above)
Comment (sibasankar-demansol) [reply server/plugin/webhook.go]: Done

**Review by mickmister — APPROVED** [review]
LGTM!

File: server/plugin/webhook.go [hunk 2]
```diff
@@ -270,6 +272,10 @@ func (p *Plugin) postPullRequestEvent(event *github.PullRequestEvent) {
 		}
 
 		if action == "opened" {
+			policy := bluemonday.StrictPolicy()
+			policy.SkipElementsContent("details")
+			newPRMessage = strings.ReplaceAll(newPRMessage, "<details>", "`(details collapsed)`<details>")
```
Comment (hanzei) [inline server/plugin/webhook.go:277]: Could you please post an example on how the output looks like?

File: go.mod [hunk 3]
```diff
@@ -8,7 +8,9 @@ require (
 	github.com/gorilla/mux v1.8.0
 	github.com/mattermost/mattermost-plugin-api v0.0.12-0.20200917103517-788aef584648
 	github.com/mattermost/mattermost-server/v5 v5.28.0
+	github.com/microcosm-cc/bluemonday v1.0.15
 	github.com/pkg/errors v0.9.1
 	github.com/stretchr/testify v1.6.1
+	golang.org/x/net v0.0.0-20210726213435-c6fcb2dbf985 // indirect
```
Comment (hanzei) [inline go.mod:14]: Nit: Remove indirect dependency

**Review by hanzei — CHANGES_REQUESTED** [review]
I think we should also apply the replacement for issue open events

File: server/plugin/webhook.go (hunk 2 above)
Comment (sibasankar-demansol) [reply server/plugin/webhook.go:277]: @hanzei  Please find the screenshot
Screenshot 2021-08-13 at 12 36 26 AM

File: go.mod (hunk 3 above)
Comment (sibasankar-demansol) [reply go.mod:14]: @hanzei tried removing it , while deploying it creates problem

**Comment by sibasankar-demansol** [comment]
> I think we should also apply the replacement for issue open events

Yes we can , please confirm if you want it

File: go.mod (hunk 3 above)
Comment (hanzei) [reply go.mod:14]: Using `go1.16.7` I can remove this line and the plugin correctly compiles. What is the error that you are seeing?

File: server/plugin/webhook.go (hunk 2 above)
Comment (hanzei) [reply server/plugin/webhook.go:277]: I'm 2/5 that it's more confusing if we replace the details with this placeholder. Does it make sense to only output the conentent of  `<summary>` here?

File: server/plugin/webhook.go [hunk 4]
```diff
@@ -380,10 +386,14 @@ func (p *Plugin) postIssueEvent(event *github.IssuesEvent) {
 		p.API.LogWarn("Failed to render template", "error", err.Error())
 		return
 	}
+	policy := bluemonday.StrictPolicy()
```
Comment (hanzei) [inline server/plugin/webhook.go:389]: Can this be DRY-ed up?

**Review by hanzei — CHANGES_REQUESTED** [review]

File: server/plugin/webhook.go (hunk 2 above)
Comment (sibasankar-demansol) [reply server/plugin/webhook.go:277]: @hanzei   Do you mean to remove the details section completely and the details collapsed completely and just show the rest of the part ?

File: go.mod (hunk 3 above)
Comment (sibasankar-demansol) [reply go.mod:14]: @hanzei  upgraded to 1.16 it works now

File: server/plugin/webhook.go (hunk 4 above)
Comment (sibasankar-demansol) [reply server/plugin/webhook.go:389]: Sure

File: server/plugin/webhook.go (hunk 2 above)
Comment (hanzei) [reply server/plugin/webhook.go:278]: I mean removing the body of the details section while keeping the `<summary` header.

Comment (hanzei) [reply server/plugin/webhook.go:278]: Does that make things more clear?

Comment (sibasankarnayak) [reply server/plugin/webhook.go:278]: @hanzei Have made the following changes
attaching the screenshot for same
```
PR created In Github
![Screenshot 2021-08-18 at 11 17 56 PM](https://user-images.githubusercontent.com/86650698/129947404-1113d226-0f7a-4e85-8000-ea9ad5307b20.png)
Trigger Message in MM GIthub bot 
![Screenshot 2021-08-18 at 11 18 11 PM](https://user-images.githubusercontent.com/86650698/129947442-41d85f26-7181-4d20-9843-8489993808e6.png)
```

Comment (hanzei) [reply server/plugin/webhook.go]: Actually, just seeing the summary does look a bit confusing. 1/5 that it's more clear to the user if we remove the whole `datails` block. What do you think?

Comment (sibasankarnayak) [reply server/plugin/webhook.go]: @hanzei  i agree this looks too ugly ,
we can simply remove the details  block and keep it clean

**Comment by sibasankarnayak** [comment]
> # [Codecov](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost) Report
> > Merging [#464](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost) ([121e5de](https://github.com/mattermost/mattermost-plugin-github/commit/121e5dec5cd37db7bb982a680962f5debef7917d)) into [master](https://codecov.io/gh/mattermost/mattermost-plugin-github/commit/bc8056eb5ce1ee9ae673c33d999a38a77e538183?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost) ([bc8056e](https://github.com/mattermost/mattermost-plugin-github/commit/bc8056eb5ce1ee9ae673c33d999a38a77e538183)) will **increase** coverage by `0.32%`.
> > The diff coverage is `0.00%`.
> 
> > ❗ Current head [121e5de](https://github.com/mattermost/mattermost-plugin-github/commit/121e5dec5cd37db7bb982a680962f5debef7917d) differs from pull request most recent head [c6948d4](https://github.com/mattermost/mattermost-plugin-github/commit/c6948d455b56947701c906863baeee7de0a89478). Consider uploading reports for the commit [c6948d4](https://github.com/mattermost/mattermost-plugin-github/commit/c6948d455b56947701c906863baeee7de0a89478) to get more accurate results
> > [Impacted file tree graph](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost)
> 
> ```diff
> @@            Coverage Diff             @@
> ##           master     #464      +/-   ##
> ==========================================
> + Coverage   19.66%   19.99%   +0.32%     
> ==========================================
>   Files          11       11              
>   Lines        2741     2991     +250     
> ==========================================
> + Hits          539      598      +59     
> - Misses       2162     2353     +191     
>   Partials       40       40              
> ```
> 
> [Impacted Files](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost)	Coverage Δ	
> [server/plugin/webhook.go](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost#diff-c2VydmVyL3BsdWdpbi93ZWJob29rLmdv)	`0.00% <0.00%> (ø)`	
> [server/plugin/utils.go](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost#diff-c2VydmVyL3BsdWdpbi91dGlscy5nbw==)	`61.53% <0.00%> (-0.40%)`	⬇️
> [server/plugin/template.go](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost#diff-c2VydmVyL3BsdWdpbi90ZW1wbGF0ZS5nbw==)	`95.10% <0.00%> (-0.20%)`	⬇️
> [server/plugin/plugin.go](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost#diff-c2VydmVyL3BsdWdpbi9wbHVnaW4uZ28=)	`4.71% <0.00%> (-0.11%)`	⬇️
> [server/plugin/manifest.go](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost#diff-c2VydmVyL3BsdWdpbi9tYW5pZmVzdC5nbw==)	`100.00% <0.00%> (ø)`	
> [server/plugin/command.go](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost#diff-c2VydmVyL3BsdWdpbi9jb21tYW5kLmdv)	`10.43% <0.00%> (+0.27%)`	⬆️
> [server/plugin/api.go](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost#diff-c2VydmVyL3BsdWdpbi9hcGkuZ28=)	`8.56% <0.00%> (+0.64%)`	⬆️
> [server/plugin/subscriptions.go](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost#diff-c2VydmVyL3BsdWdpbi9zdWJzY3JpcHRpb25zLmdv)	`10.71% <0.00%> (+0.92%)`	⬆️
> [server/plugin/permalinks.go](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost#diff-c2VydmVyL3BsdWdpbi9wZXJtYWxpbmtzLmdv)	`70.37% <0.00%> (+1.94%)`	⬆️
> ... and [2 more](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464/diff?src=pr&el=tree-more&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost)		
> [Continue to review full report at Codecov](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost).
> 
> > **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost)
> > `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> > Powered by [Codecov](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464?src=pr&el=footer&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost). Last update [bc8056e...c6948d4](https://codecov.io/gh/mattermost/mattermost-plugin-github/pull/464?src=pr&el=lastupdated&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mattermost).

/check-cla

File: server/plugin/webhook.go [hunk 5]
```diff
@@ -380,6 +386,8 @@ func (p *Plugin) postIssueEvent(event *github.IssuesEvent) {
 		p.API.LogWarn("Failed to render template", "error", err.Error())
 		return
 	}
+	renderedMessage = p.sanitizeDescription(renderedMessage)
```
Comment (hanzei) [inline server/plugin/webhook.go:389]: What about PRs?

**Review by hanzei — CHANGES_REQUESTED** [review]

File: server/plugin/webhook.go (hunk 5 above)
Comment (sibasankarnayak) [reply server/plugin/webhook.go:389]: We have done it for both opened issue and Open PR

Comment (sibasankarnayak) [reply server/plugin/webhook.go:389]: now it looks like this 

Screenshot 2021-08-19 at 5 57 12 PM

Comment (hanzei) [reply server/plugin/webhook.go:390]: It seems like the markdown also get erased. Can you please take a look what is happening there?

Comment (sibasankarnayak) [reply server/plugin/webhook.go:390]: @hanzei  tried to figure out but i didn't find any difference  tried the same desc with the code in master to compare 
it is same

Comment (hanzei) [reply server/plugin/webhook.go:390]: That's surprising. Could you also please test the latest release? On the community server markdown renders fine:

Comment (sibasankarnayak) [reply server/plugin/webhook.go:390]: @hanzei 
i updated my code for mm-server , mm-webapp , mm-plugin-github
and created a PR and the Message looks like this

Screenshot 2021-08-20 at 9 45 00 AM

Screenshot 2021-08-20 at 9 44 39 AM

Comment (hanzei) [reply server/plugin/webhook.go:390]: What is the content of the original GitHub message? Does it contain md or html?

Comment (sibasankarnayak) [reply server/plugin/webhook.go:390]: @hanzei  i tried with this
```
Bumps <a href="https://github.com/npm/hosted-git-info">hosted-git-info</a> from 2.8.8 to 2.8.9. <strong>This update includes a security fix.</strong>
```

File: server/plugin/webhook.go [hunk 6]
```diff
@@ -270,7 +272,8 @@ func (p *Plugin) postPullRequestEvent(event *github.PullRequestEvent) {
 		}
 
 		if action == "opened" {
-			post.Message = newPRMessage
+			p.API.LogWarn("logs", "PR", p.sanitizeDescription(newPRMessage))
```
Comment (hanzei) [inline server/plugin/webhook.go:275]: Left over from debugging?

File: server/plugin/webhook.go (hunk 5 above)
Comment (hanzei) [reply server/plugin/webhook.go:390]: Okay, that's no md input. I that case this PR didn't introduce a regression. LGTM :+1:

**Review by hanzei — CHANGES_REQUESTED** [review]
Almost ready to merge. There seems to be one debug log message remaining.

File: server/plugin/webhook.go (hunk 6 above)
Comment (sibasankarnayak) [reply server/plugin/webhook.go]: @hanzei  sorry my bad , removed it

**Review by hanzei — APPROVED** [review]
Awesome :tada:

**Pull request merged**